erizo/src/erizo/WebRtcConnection.cpp

```diff
@@ -98,41 +98,63 @@ void WebRtcConnection::close() {
 }
 
 bool WebRtcConnection::init() {
-    maybeNotifyWebRtcConnectionEvent(global_state_, "");
-    return true;
+  maybeNotifyWebRtcConnectionEvent(global_state_, "");
+  return true;
 }
 
-bool WebRtcConnection::createOffer(bool video_enabled, bool audioEnabled, bool bundle) {
+std::shared_ptr<std::promise<void>> WebRtcConnection::createOffer(bool video_enabled, bool audio_enabled, bool bundle) {
+  return asyncTask([video_enabled, audio_enabled, bundle] (std::shared_ptr<WebRtcConnection> connection) {
+    connection->createOfferSync(video_enabled, audio_enabled, bundle);
+  });
+}
+
+bool WebRtcConnection::createOfferSync(bool video_enabled, bool audio_enabled, bool bundle) {
   boost::mutex::scoped_lock lock(update_state_mutex_);
   bundle_ = bundle;
   video_enabled_ = video_enabled;
-  audio_enabled_ = audioEnabled;
+  audio_enabled_ = audio_enabled;
   local_sdp_->createOfferSdp(video_enabled_, audio_enabled_, bundle_);
+
   local_sdp_->dtlsRole = ACTPASS;
+  if (local_sdp_->internal_dtls_role == ACTPASS) {
+    local_sdp_->internal_dtls_role = PASSIVE;
+  }
 
-  ELOG_DEBUG("%s message: Creating sdp offer, isBundle: %d", toLog(), bundle_);
+  ELOG_DEBUG("%s message: Creating sdp offer, isBundle: %d, setup: %d",
+    toLog(), bundle_, local_sdp_->internal_dtls_role);
 
-  if (video_enabled_) {
-    forEachMediaStream([this] (const std::shared_ptr<MediaStream> &media_stream) {
+  forEachMediaStream([this] (const std::shared_ptr<MediaStream> &media_stream) {
+    if (!media_stream->isReady() || media_stream->isPublisher()) {
```

Comment:
I'm not following exactly what we're trying to accomplish with `ready_ ` in `MediaStream`.
It looks like it is always initialized with `forced=true` so it's always ready right now, is this WIP?. Is the idea that there will be mediaStreams that we have to be sure a remote Sdp has been passed and processed? In that case I would change the name of the `force` parameter to something like `doNotWaitForRemoteSdp` or something like that.

Also, what is the effect of the MediaStream not being ready here? It is going to be ignored in the SDP, are we going to handle that differently in the future?

Reply:
yes, you're tight, the idea is that we won't add streams that are not ready to the SDP. Previously, we added them if they were initialized. They can be ready if we receive a remote offer or if we sent a local offer, so it shouldn't change the way it used to worked for previous cases.
I'll change the parameter name to `doNotWaitForRemoteSdp` as you propose because it's more descriptive.